src/sso/src/main.rs

```diff
@@ -0,0 +1,316 @@
+/*
+   Unix Azure Entra ID implementation
+   Copyright (C) David Mulder <dmulder@samba.org> 2024
+
+   This program is free software; you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation; either version 3 of the License, or
+   (at your option) any later version.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+use serde_json::json;
+mod broker;
+use broker::BrokerClient;
+use clap::{CommandFactory, Parser, Subcommand};
+use serde::Serialize;
+use serde_json::Value;
+use std::env;
+use std::error::Error;
+use std::io::{self, Read, Write};
+use uuid::Uuid;
+
+const SSO_URL_DEFAULT: &str = "https://login.microsoftonline.com/";
+const EDGE_BROWSER_CLIENT_ID: &str = "d7b530a4-7680-4c23-a8bf-c52c121d2e87";
+const GRAPH_SCOPES: [&str; 1] = ["https://graph.microsoft.com/.default"];
+
+macro_rules! block_read {
+    ($buf:expr) => {{
+        loop {
+            match io::stdin().read_exact(&mut $buf) {
+                Ok(_) => break,
+                Err(ref e) if e.kind() == io::ErrorKind::UnexpectedEof => {
+                    continue;
+                }
+                Err(e) => {
+                    return Err(Box::new(e));
+                }
+            }
+        }
+    }};
+}
+
+struct NativeMessaging;
+
+impl NativeMessaging {
+    fn get_message() -> Result<serde_json::Value, Box<dyn Error>> {
+        let mut raw_length = [0u8; 4];
+        block_read!(raw_length);
+        let message_length = u32::from_le_bytes(raw_length) as usize;
+
+        let mut buffer = vec![0u8; message_length];
+        io::stdin().read_exact(&mut buffer)?;
+
+        let message_str = String::from_utf8(buffer)?;
+        Ok(serde_json::from_str(&message_str)?)
+    }
+
+    fn encode_message<T: Serialize>(message_content: &T) -> Result<Vec<u8>, Box<dyn Error>> {
+        let encoded_content = serde_json::to_vec(message_content)?;
+        let mut result = Vec::new();
+        result.extend(&(encoded_content.len() as u32).to_le_bytes());
```

Comment:
I think the specification says native endianness. The python app uses `struct.pack("@I", ...)` here, so `to_ne_bytes` should be correct.

Reply:
Well, I've tried modifying that line and using native endianness, but it still fails the same way.
![Screenshot From 2025-01-07 14-26-45](https://github.com/user-attachments/assets/e6e313a8-ed03-47a0-977b-6a21d7b54a6c)

Reply:
You could create a breakpoint in the JS background task, there should be a function named `on_native_message`. From there you could inspect the native message sent by your implementation and see if something is wrong with it.